qboost/demo.ipynb

```diff
@@ -396,8 +463,228 @@
+    "print('=======================================')\n",
+    "# Decision Tree\n",
+    "print('Decision Tree: ')\n",
+    "clf1 = Decision_Tree(X_train, y_train, X_test, y_test) \n",
+    "print('---------------------------------------')\n",
+    "# Random Forest\n",
+    "print('Random Forest: ')\n",
+    "clf2 = Random_Forest(X_train, y_train, X_test, y_test) \n",
+    "print('---------------------------------------')\n",
+    "# AdaBoost\n",
+    "print('AdaBoost: ')\n",
+    "clf3 = AdaBoost(X_train, y_train, X_test, y_test) \n",
+    "print('---------------------------------------')\n",
+    "# QBoost\n",
+    "print('QBoost: ')\n",
+    "clf4 = QBoost(X_train, y_train, X_test, y_test) \n",
+    "print('---------------------------------------')\n",
+    "# QBoostPlus\n",
+    "print('QBoostPlus: ')\n",
+    "clf5 = QBoostPlus(X_train, y_train, X_test, y_test, clf1, clf2, clf3, clf4) \n",
+    "print('=======================================')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# A Few More Words on Ensemble Methods\n",
+    "\n",
+    "Ensemble methods build a strong classifier (an improved model) by combining weak classifiers with the goal of:\n",
+    "\n",
+    "* decreasing variance (bagging)\n",
+    "* decreasing bias (boosting)\n",
+    "* improving prediction (voting)\n",
+    "\n",
+    "![Boosting Algorithm](images/boosting.jpg)\n",
+    "\n",
+    "### Bagging, Boosting, and Voting\n",
+    "\n",
+    "The ensemble method produces new training data sets by random sampling with replacement from the original set. In _bagging_, any element has the same probability to appear in a new dataset; in _boosting_, data elements are weighted before they are collected in the new dataset. Another distinction is that bagging is parallelizable but boosting has to be executed sequentially. You can learn more about the differences between these methods here: https://quantdare.com/what-is-the-difference-between-bagging-and-boosting/.\n",
+    "\n",
+    "Voting operates on labels only. Unlike boosting, the aggeragated classification performance is not used to further polish each weak classifier. Voting has two typical requirements of its collection of  weak classifiers: that there be __many__ and that they be __diverse__.  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Under the Hood:  Comparing AdaBoost and QBoost\n",
+    "\n",
+    "### AdaBoost\n",
+    "AdaBoost combines a number of $N$ weak classifiers into a strong one as\n",
+    "$$C(x) = sign\\left(\\sum_i^N w_i c_i(x)\\right),$$\n",
+    "with $c_i(x) \\in [-1, +1]$ being the $i$-th weak classifier:\n",
+    "\n",
+    "$$c_i(x) = sign(w'*x + b)$$\n",
+    "\n",
+    "The loss function of AdaBoost is defined as\n",
+    "$$\n",
+    "L = \\sum_{n=1}^N \\exp\\left\\{ - y_n \\sum_{s=1}^S w_sc_k(x_n)\\right\\}.\n",
```

Comment:
Perhaps a sentence explaining what the variables all mean and the dimensions of each variable? It's just a little difficult to understand the equation when you aren't certain about what each of the variables represent.

ex. something like "`x_i` is a n-element vector, where each of its elements represent a pixel/feature in the MNIST image"